core/modules/xrdsvc/StreamBuffer.h

```diff
@@ -0,0 +1,96 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2014-2018LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_XRDSVC_STREAMBUFFER_H
+#define LSST_QSERV_XRDSVC_STREAMBUFFER_H
+
+// System headers
+#include <atomic>
+#include <condition_variable>
+#include <deque>
+#include <mutex>
+#include <string>
+
+// qserv headers
+#include "util/InstanceCount.h"
+
+// Third-party headers
+#include "XrdSsi/XrdSsiErrInfo.hh" // required by XrdSsiStream
+#include "XrdSsi/XrdSsiStream.hh"
+
+namespace lsst {
+namespace qserv {
+namespace xrdsvc {
+
+
+/// StreamBuffer is a single use buffer for transferring data packets
+/// to XrdSsi.
+/// Its notable feature is the Recycle() function, which XrdSsi will
+/// promptly call when it no longer needs the buffer.
+class StreamBuffer : public XrdSsiStream::Buffer {
+public:
+    using Ptr = std::shared_ptr<StreamBuffer>;
+
+    // Copying this would be very confusing for something waiting for Recycle().
+    StreamBuffer() = delete;
+    StreamBuffer(StreamBuffer const&) = delete;
+    StreamBuffer& operator=(StreamBuffer const&) = delete;
+
+    /// Factory function, because this should be able to delete itself when Recycle() is called.
+    //  The constructor uses move to avoid copying the string.
+    static StreamBuffer::Ptr createWithMove(std::string &input);
```

Comment:
I'd probably do `string&& input` here so that temporary strings can be passed too.